File: 2017/round3/googlements.go
package main

import (
	"fmt"
	"reflect"
	"sort"
	"strconv"
)

func main() {
	var nTests int
	fmt.Scanf("%d\n", &nTests)
	for i := 0; i < nTests; i++ {
		var val string
		fmt.Scanf("%s\n", &val)
		fmt.Println(val)
		vs := val2IntArr(val)
		count := solve(vs)
		fmt.Printf("Case #%d: %d\n", i+1, count)
	}
}

func val2IntArr(val string) []int {
	result := []int{}
	for _, c := range val {
		i, _ := strconv.Atoi(string(c))
		result = append(result, i)
	}
	return result
}

func solve(vs []int) int {
	if !isValidGooglement(vs) {
		return 1
	}

	count := 0
	perms := permutations(getAncestorNums(vs))
	for _, perm := range perms {
		if reflect.DeepEqual(perm, vs) {
			continue
		}
		c := solve(perm)
		count += c
	}

	return count
}

func getAncestorNums(vs []int) []int {
	var result []int
	for num, count := range vs {
		for i := 0; i < count; i++ {
			result = append(result, num)
		}
	}
	return result
}

func isValidGooglement(vs []int) bool {
	sum := 0
	for _, v := range vs {
		sum += v
	}
	if sum > len(vs) {
		return false
	}
	return true
}

func permutations(arr []int) [][]int {
	var helper func([]int, int)
	res := [][]int{}

	helper = func(arr []int, n int) {
		if n == 1 {
			tmp := make([]int, len(arr))
			copy(tmp, arr)
			res = append(res, tmp)
		} else {
			for i := 0; i < n; i++ {
				helper(arr, n-1)
				if n%2 == 1 {
					tmp := arr[i]
					arr[i] = arr[n-1]
					arr[n-1] = tmp
				} else {
					tmp := arr[0]
					arr[0] = arr[n-1]
					arr[n-1] = tmp
				}
			}
		}
	}
	helper(arr, len(arr))
	return res
}

File: 2017/round3/googlements.py
import itertools
import math

def solve(vals):
    count = 1
    if not is_valid_googlement(vals):
        return count

    anums = generate_ancestor_nums(vals)
    if not is_valid_googlement(anums):
        return count + just_count(vals)

    for perm in set(itertools.permutations(anums)):
        plist = list(perm)
        if not vals == plist:
            count += solve(plist)
    return count

def is_valid_googlement(vals):
    return len(vals) >= sum(vals)

def just_count(vals):
    f = math.factorial
    rem = len(vals)
    count = 1
    for i, val in enumerate(vals):
        count *= f(rem) / f(val) / f(rem - val)
        rem -= val
    return count

def generate_ancestor_nums(vals):
    result = []
    for i, val in enumerate(vals):
        for j in range(val):
            result.append(i + 1)
    s = sum(vals)
    l = len(vals)
    if l > s:
        for i in range(l - s):
            result.append(0)
    return result

T = int(raw_input())

for t in range(T):
    vals = [int(d) for d in raw_input()]
    count = solve(vals)
    print("Case #%d: %d" % (t + 1, count))
